### PR TITLE
[FSDP] Start to generalize modules to ignore for mixed precision

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102231
* #102043
* #102040
* __->__ #102010

The main use case here is that folks would like to ignore layer norm for mixed precision. This can now be enabled with:

```
mp_config = MixedPrecision(
            param_dtype=torch.float16,
            reduce_dtype=torch.float16,
            buffer_dtype=torch.float16,
            _mixed_precision_module_classes_to_ignore=[_BatchNorm, nn.LayerNorm],
        )
```

This is done by classes of types in `_mixed_precision_module_classes_to_ignore` being wrapped in their own FSDP unit with mixed preicsion disabled. This is only enabled for auto wrapping.

We also add module pre and post hooks to cast / downcast inputs to the appropriate full precision.

Differential Revision: [D46079957](https://our.internmc.facebook.com/intern/diff/D46079957/)

### CODE REVIEW

File: torch/distributed/fsdp/_utils.py [hunk 1]
```diff
@@ -2,13 +2,47 @@
...
     return any(isinstance(mod, _BatchNorm) for mod in module.modules())
 
 
+def _contains_module(root, module_to_check):
+    return any(isinstance(mod, module_to_check) for mod in root.modules())
+
+def _override_module_mixed_precision(root, module_to_override, wrap_override_dict={"mixed_precision": None}):
+    for mod in root.modules():
+        if isinstance(mod, module_to_override):
+            mod._wrap_overrides = wrap_override_dict  # type: ignore[assignment]
```
Comment (awgu) [inline torch/distributed/fsdp/_utils.py:22]: This looks unused?

File: torch/distributed/fsdp/_utils.py [hunk 2]
```diff
@@ -2,13 +2,47 @@
...
+
+        if isinstance(mod, LayerNorm): # TODO: generalize this logic if additional types need to be supported
```
Comment (awgu) [inline torch/distributed/fsdp/_utils.py:24]: I think this should be replaced with `if isinstance(mod, module_to_override):`?

File: torch/distributed/fsdp/api.py [hunk 3]
```diff
@@ -133,6 +134,11 @@ class MixedPrecision:
             arguments and keyword arguments to ``param_dtype`` for the root FSDP instance.
             It takes precedence over ``cast_forward_inputs`` for the root FSDP instance.
             (Default: ``True``)
+        _mixed_precision_module_classes_to_ignore: (Sequence[type]): Module classes to ignore
+            for mixed precision. This will make the specified nn.Module types ignore mixed precision,
```
Comment (awgu) [inline torch/distributed/fsdp/api.py:138]: I think this comment should mention that this setting only affects auto wrapping.

File: torch/distributed/fsdp/_wrap_utils.py [hunk 4]
```diff
@@ -57,10 +62,21 @@ def _auto_wrap(
                 "if using an `auto_wrap_policy`"
             )
     mixed_precision = fsdp_kwargs["mixed_precision"]
-    if mixed_precision is not None and _contains_batchnorm(root_module):
-        _override_batchnorm_mixed_precision(root_module)
+    if mixed_precision is not None and any(
+        _contains_module(root_module, module_to_override)
+        for module_to_override in mixed_precision._mixed_precision_module_classes_to_ignore
+    ):
+        for mp_module_to_override in mixed_precision._mixed_precision_module_classes_to_ignore:
```
Comment (awgu) [inline torch/distributed/fsdp/_wrap_utils.py:69]: Is one of these for loops redundant?

File: torch/distributed/fsdp/_utils.py [hunk 5]
```diff
@@ -2,13 +2,47 @@
...
+            def ln_forward_pre_hook(module, args):
+                return _apply_to_tensors(partial(cast_fn, torch.float32), args)
+
+            def ln_forward_post_hook(module, args, output):
+                nonlocal old_dtype
+                if old_dtype is not None:
+                    return _apply_to_tensors(partial(cast_fn, old_dtype), output)
+
+            mod.register_forward_pre_hook(ln_forward_pre_hook)
+            mod.register_forward_hook(ln_forward_post_hook)
```
Comment (awgu) [inline torch/distributed/fsdp/_utils.py:44]: Do we strictly need these hooks for correctness? If so, could we make sure we have unit tests that exercise that?

**Review by awgu — COMMENTED** [review]
I think the general idea makes sense to me, but there is some work to clean up the PR before landing. I am mostly interested in what is the minimal set of changes as there seems to be some redundancy.

It seems like there are two things going on: the added hooks for non-FSDP submodules within an FSDP's tree and also auto-wrapping specified module types in their own FSDP instances. The PR summary focuses on the former -- should we also mention the latter if we include it in this PR?

File: torch/distributed/fsdp/_utils.py (hunk 1 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/_utils.py:22]: wrap_overrides is populated here and then used in auto wrapping: https://fburl.com/code/7ypnkd83

File: torch/distributed/fsdp/_utils.py (hunk 2 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/_utils.py:24]: I was hardcoding this to LN for now. LN would crash if it got fp16 inputs and was running in fp32, but for some reason batchnorm doesn't need this. Maybe we should just always cast up to fp32?

Comment (awgu) [reply torch/distributed/fsdp/_utils.py:24]: Oh I see. If batch norm kernels have the casting logic, then maybe layer norm should as well? It might be good to check on why there is that difference before adding logic at the FSDP level.

File: torch/distributed/fsdp/_utils.py [hunk 6]
```diff
@@ -2,13 +2,47 @@
...
+    return any(isinstance(mod, module_to_check) for mod in root.modules())
+
+def _override_module_mixed_precision(root, module_to_override, wrap_override_dict={"mixed_precision": None}):
+    for mod in root.modules():
+        if isinstance(mod, module_to_override):
+            mod._wrap_overrides = wrap_override_dict  # type: ignore[assignment]
+
+        if isinstance(mod, LayerNorm): # TODO: generalize this logic if additional types need to be supported
+
+            old_dtype = None
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/distributed/fsdp/_utils.py:26]: Wonder what we can do here. The input -> output mapping for a module is not necessarily 1:1. We're basically saving an `old_dtype` so that we cast the output back to whatever the input precision was before forcing to fp32. But agreed there can be multiple input precisions.

Maybe we don't cast back at all (so we don't need a post hook or old dtype), and rely on users doing `cast_forward_inputs` configuring correctly to ensure dtypes into subsequent mixed precision modules are correct?

File: torch/distributed/fsdp/_utils.py (hunk 5 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/_utils.py:44]: Yeah, with the LN added to the test, it would crash without the pre hook. The necesity of the post hook depends on the `old_dtype` discussion above

File: torch/distributed/fsdp/_utils.py [hunk 7]
```diff
@@ -2,13 +2,47 @@
...
+
+            def ln_forward_pre_hook(module, args):
+                return _apply_to_tensors(partial(cast_fn, torch.float32), args)
+
+            def ln_forward_post_hook(module, args, output):
+                nonlocal old_dtype
+                if old_dtype is not None:
+                    return _apply_to_tensors(partial(cast_fn, old_dtype), output)
+
+            mod.register_forward_pre_hook(ln_forward_pre_hook)
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/distributed/fsdp/_utils.py:43]: done

File: torch/distributed/fsdp/_wrap_utils.py (hunk 4 above)
Comment (rohan-varma) [reply torch/distributed/fsdp/_wrap_utils.py:69]: yeah

**Review by awgu — COMMENTED** [review]
I sent a message on chat to discuss more about the hooks casting.

**Review by awgu — APPROVED** [review]
This makes sense to me! It will be good to think more about the hooks design as we try this out in more use cases.

**Comment by awgu** [comment]
Lint error looks real :/
